### PR TITLE
ENH: Move symiirorder1/2, cspline2d, qspline2d and spline_filter to use sosfilt/lfilter

### PR DESCRIPTION
**Pull request by andfoy**
#### What does this implement/fix?
This PR deletes the following redundant linear filter implementations: `*_IIR_forback1`, `*_IIR_forback1` and replace them to calls to `lfilter`/`sosfilt` in order to have a single filter implementation. As a result, several C functions will be removed.

#### Additional information

This is related to the GPU implementations of 2D B-Splines in CuPy: https://github.com/cupy/cupy/pull/7721

### CODE REVIEW

**Comment by ev-br** [comment]
Needs a rebase

File: scipy/signal/_bsplines.py
Comment (ev-br) [inline scipy/signal/_bsplines.py:439]: This was https://github.com/scipy/scipy/pull/18926/files#diff-1bcc32e1c74a323eb44f3afc3cd2a9d3b4b64c32654b9e775c7969daa33e5973L76 and simiar for doubles, correct? Then for backwards compat could you please keep `if precision < -1 or precision > 1`.

While at it, might change `dtype.char` to `if signal.dtype in [np.foat32, np.complex64]`.

File: scipy/signal/_splinemodule.c
Comment (ev-br) [inline scipy/signal/_splinemodule.c:89]: Could you comment why is this no longer possible or needed?

File: scipy/signal/tests/test_splines.py
Comment (ev-br) [inline scipy/signal/tests/test_splines.py:94]: this differs from the symiirorder1 implementation?

File: scipy/signal/tests/test_splines.py
Comment (ev-br) [inline scipy/signal/tests/test_splines.py:86]: Would be nice to add a test with integer inputs, just to make sure backwards compat is OK:

```
In [16]: s = np.where(np.arange(100) %2, -1, 1)

In [17]: signal.symiirorder1(s.astype(float), 0.5, 0.5) == signal.symiirorder1(s, 0.5, 0
    ...: .5)
Out[17]: 
array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

In [18]: scipy.__version__
Out[18]: '1.11.1'
```

File: scipy/signal/_splines.py
Comment (ev-br) [inline scipy/signal/_splines.py:130]: Is this correct for `input`s of integer dtype?

**Review by ev-br — COMMENTED** [review]
A few minor comments / questions.

File: scipy/signal/_splinemodule.c
Comment (andfoy) [reply scipy/signal/_splinemodule.c:89]: This is no longer needed since the function does not call malloc anymore

File: scipy/signal/tests/test_splines.py
Comment (andfoy) [reply scipy/signal/tests/test_splines.py:94]: The precision default values are embedded in the C module function definition, these are explicit here since the actual values are required to compute the closed form equations.

File: scipy/signal/_splines.py
Comment (andfoy) [reply scipy/signal/_splines.py:130]: Let me check

**Comment by andfoy** [comment]
> python side: any reason to not merge _bsplines.py and _splines.py?

The C module is called _spline, so I thought it was appropriate to move the functions to a file called _splines, as it was done on CuPy. We can fuse the files if necessary.

**Review by ev-br — COMMENTED** [review]
A few more asks for code comments. 

As discussed offline, we are likely to punt on larger items I was asking for, e.g. fixing the implementation for complex and removing the `__GNUC__` guards can defer to a follow-up.

File: scipy/signal/_splines.py
Comment (ev-br) [inline scipy/signal/_splines.py:53]: So this relies on numpy propotining `(int, float32) -> float64`, it seems (which works on both numpy < 2 and 2.0.0rc1). Am I missing a reason to spell it like this instead of just `signal.astype(float)`?

**Review by ev-br — APPROVED** [review]

**Comment by ev-br** [comment]
LGTM modulo a minor comment, which can be done here or in a follow-up.

The main changes in the PR are great IMO:
- actual filtering is delegated to `lfilter/sosfilt` instead of custom implementations;
- test coverage is better.

Several further improvements are possible:
- consolidate python and C implementation files (it's not very clear what is in `_splines` and what is in `_bspline_utils`), maybe rename to `_spline_filters` or some such.
- improve handling of complex types, remove GNUC-specific pragmas.
- remove manual templating to C++ templates.

Am happy to push these to a follow-up though, this PR is an improvement as is.

Am going to keep this open for a bit in case there are further comments/reviews.

File: scipy/signal/_splines.py
Comment (andfoy) [reply scipy/signal/_splines.py:53]: Is `float` an alias to the smallest floating type that can represent the values?

Comment (ev-br) [reply scipy/signal/_splines.py:53]: No, `s.astype(float)` is an alias to `s.astype(np.float64)`. Which is what the `promote_type` stanza above resolves to:

```
In [1]: import numpy as np

In [2]: s = np.where(np.arange(100) %2, -1, 1)

In [3]: s.dtype
Out[3]: dtype('int64')

In [4]: np.promote_types(s.dtype, np.float32)
Out[4]: dtype('float64')
```

Which is correct from the backwards compat POV, since

```
In [6]: symiirorder1(s, 0.5, 0.5).dtype
Out[6]: dtype('float64')

In [7]: import scipy

In [8]: scipy.__version__
Out[8]: '1.7.1'

```

**Comment by ev-br** [comment]
Poking around some more, and here's a naive question. If I compare the numeric values to previous scipy versions, `symiirorder2` seems to require rather large tolerance (~2e-6 for precision =1e-10). Moreover, the differences are seen at the end of the signal, and seem to be only weakly dependent on `precision`.  

The test is https://github.com/ev-br/scipy/commit/b5aeddf47dde7a6c406896f5d72f79d8595b62cc; if I tighten the tolerance for symiir2, the failures are

```
========================================================== FAILURES ==========================================================
__________________________________________ TestSymIIR.test_symiir2_values[float32] ___________________________________________
scipy/signal/tests/test_splines.py:338: in test_symiir2_values
    assert_allclose(res, exp_res, atol=atol)
        atol       = 1e-15
        dtyp       = <class 'numpy.float32'>
        exp_res    = array([0.26572609, 0.53408018, 0.51032696, 0.72115829, 0.69486885,
       0.3649055 , 0.37349478, 0.74165032, 0.89718521, 0.80582483,
       0.46758053, 0.51898709, 0.65025605, 0.65394321, 0.45273595,
       0.53539183])
        res        = array([0.26572606, 0.5340801 , 0.5103269 , 0.72115827, 0.69486886,
       0.36490548, 0.37349477, 0.7416503 , 0.8971852 , 0.80582476,
       0.46758053, 0.51898706, 0.65025604, 0.65394306, 0.45273548,
       0.5353913 ], dtype=float32)
        rng        = RandomState(MT19937) at 0x7F2110FE1440
        s          = array([0.19151945, 0.62210876, 0.43772775, 0.7853586 , 0.77997583,
       0.2725926 , 0.27646425, 0.8018722 , 0.95813936, 0.87593263,
       0.35781726, 0.5009951 , 0.6834629 , 0.71270204, 0.37025076,
       0.5611962 ], dtype=float32)
        self       = <scipy.signal.tests.test_splines.TestSymIIR object at 0x7f2110beb910>
/home/br/mambaforge/envs/scipy-dev/lib/python3.10/contextlib.py:79: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=1e-15
E   
E   Mismatched elements: 5 / 16 (31.2%)
E   Max absolute difference: 5.58885651e-07
E   Max relative difference: 1.04388154e-06
E    x: array([0.265726, 0.53408 , 0.510327, 0.721158, 0.694869, 0.364905,
E          0.373495, 0.74165 , 0.897185, 0.805825, 0.467581, 0.518987,
E          0.650256, 0.653943, 0.452735, 0.535391], dtype=float32)
E    y: array([0.265726, 0.53408 , 0.510327, 0.721158, 0.694869, 0.364905,
E          0.373495, 0.74165 , 0.897185, 0.805825, 0.467581, 0.518987,
E          0.650256, 0.653943, 0.452736, 0.535392])
        args       = (<function assert_allclose.<locals>.compare at 0x7f2110c63520>, array([0.26572606, 0.5340801 , 0.5103269 , 0.72115827,...165032, 0.89718521, 0.80582483,
       0.46758053, 0.51898709, 0.65025605, 0.65394321, 0.45273595,
       0.53539183]))
        func       = <function assert_array_compare at 0x7f211472d900>
        kwds       = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-07, atol=1e-15', 'verbose': True}
        self       = <contextlib._GeneratorContextManager object at 0x7f211471a920>
__________________________________________ TestSymIIR.test_symiir2_values[float64] ___________________________________________
scipy/signal/tests/test_splines.py:338: in test_symiir2_values
    assert_allclose(res, exp_res, atol=atol)
        atol       = 1e-15
        dtyp       = <class 'numpy.float64'>
        exp_res    = array([0.26572609, 0.53408018, 0.51032696, 0.72115829, 0.69486885,
       0.3649055 , 0.37349478, 0.74165032, 0.89718521, 0.80582483,
       0.46758053, 0.51898709, 0.65025605, 0.65394321, 0.45273595,
       0.53539183])
        res        = array([0.26572609, 0.53408018, 0.51032696, 0.72115829, 0.69486885,
       0.3649055 , 0.37349478, 0.74165032, 0.89718521, 0.80582483,
       0.46758053, 0.51898709, 0.65025604, 0.65394314, 0.45273554,
       0.53539127])
        rng        = RandomState(MT19937) at 0x7F2110FE2D40
        s          = array([0.19151945, 0.62210877, 0.43772774, 0.78535858, 0.77997581,
       0.27259261, 0.27646426, 0.80187218, 0.95813935, 0.87593263,
       0.35781727, 0.50099513, 0.68346294, 0.71270203, 0.37025075,
       0.56119619])
        self       = <scipy.signal.tests.test_splines.TestSymIIR object at 0x7f2110c07a90>
/home/br/mambaforge/envs/scipy-dev/lib/python3.10/contextlib.py:79: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=1e-15
E   
E   Mismatched elements: 3 / 16 (18.8%)
E   Max absolute difference: 5.598832e-07
E   Max relative difference: 1.04574476e-06
E    x: array([0.265726, 0.53408 , 0.510327, 0.721158, 0.694869, 0.364905,
E          0.373495, 0.74165 , 0.897185, 0.805825, 0.467581, 0.518987,
E          0.650256, 0.653943, 0.452736, 0.535391])
E    y: array([0.265726, 0.53408 , 0.510327, 0.721158, 0.694869, 0.364905,
E          0.373495, 0.74165 , 0.897185, 0.805825, 0.467581, 0.518987,
E          0.650256, 0.653943, 0.452736, 0.535392])
```

**Comment by ev-br** [comment]
No further comments, the patch still LGTM, let's merge this. Thank you @andfoy !

**Pull request merged**
